### PR TITLE
Realtime Audio Streaming

### PR DESCRIPTION
**Pull request by shushantsharan**
**Design**

Screenshot 2021-04-13 at 1 51 00 PM

**Detailed Documentation**

https://github.com/voiceip/oreka/wiki/Live-Streaming

### CODE REVIEW

File: orkaudio/filters/LiveStream/LiveStream.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+	char sound_type = 1;
+
+	timestamp += 160; //Timestamp increment = clock frequency/frame rate
+					  //160 byte payload of G.711 has a packetization interval of 20 ms
+					  //For 1 second, there will be 1000ms / 20ms = 50 frames
+					  //Audio RTP packet timestamp incremental value = 8kHz / 50 = 8000Hz / 50 = 160
+
+	if (outputDetails.m_channel == 1)
+	{
+		bufferQueue.push(inputBuffer);
```
Comment (shushantsharan) [inline orkaudio/filters/LiveStream/LiveStream.cpp:126]: First channel packets are stored in a queue and when we receive a packet from the second channel, a packet from the queue is popped and processed together

File: orkbasecxx/CapturePluginProxy.h [hunk 2]
```diff
@@ -27,6 +27,9 @@ class DLL_IMPORT_EXPORT_ORKBASE CapturePluginProxy
 	void Run();
 	void Shutdown();
 	void StartCapture(CStdString& party, CStdString& orkuid, CStdString& nativecallid, CStdString& side);
+	void StartStream(CStdString& party, CStdString& orkuid, CStdString& nativecallid);
+	void EndStream(CStdString& party, CStdString& orkuid, CStdString& nativecallid);
+	CStdString GetStream();
```
Comment (kingster) [inline orkbasecxx/CapturePluginProxy.h:32]: IMO this should be exposed by the Filter and not Proxy. It can implement the `StartCapture`, `StopCapture` to offer the same. Please refactor it to be part of the `LiveStreamFilter` folder

File: orkaudio/filters/LiveStream/LiveStream.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "LogManager.h"
+#include "Filter.h"
+#include <log4cxx/logger.h>
+#include "Utils.h"
+#include "srs_librtmp.h"
+#include <queue>
+
+
+class DLL_IMPORT_EXPORT_ORKBASE LiveStreamFilter : public Filter
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStream.h:20]: Rename file to match ClassName `LiveStreamFilter` .

File: orkaudio/filters/LiveStream/LiveStream.cpp [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+static log4cxx::LoggerPtr s_log = log4cxx::Logger::getLogger("plugin.livestream");
+
+template<class T>
+std::string toString(const T &value) {
+    std::ostringstream os;
+    os << value;
+    return os.str();
+}
+
+LiveStreamFilter::LiveStreamFilter()
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStream.cpp:30]: Rename file to match className.

File: orkaudio/audiocaptureplugins/voip/VoIpSession.cpp [hunk 5]
```diff
@@ -4503,6 +4504,113 @@ CStdString VoIpSessions::StartCaptureNativeCallId(CStdString& nativecallid, CStd
...
+	{
+		session = pair->second;
+
+		if(session->NativeCallIdMatches(nativecallid))
+		{
+			session->m_keepRtp = true;
+			found = true;
+			orkUid = session->GetOrkUid();
+		}
+	}
```
Comment (kingster) [inline orkaudio/audiocaptureplugins/voip/VoIpSession.cpp:4565]: Duplicate code in Start/End function, please refactor

File: orkaudio/filters/LiveStream/LiveStream.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+
+  private:
+    AudioChunkRef m_outputAudioChunk;
+    bool m_initialized;
+    CStdString m_callId;
+    bool status = true;
+    srs_rtmp_t rtmp = NULL;
+    u_int32_t timestamp = 0;
+    std::queue<char*> bufferQueue;
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStream.h:45]: Maybe add this to limit the maxSize?

File: orkaudio/audiocaptureplugins/voip/VoIp.cpp [hunk 7]
```diff
@@ -1951,6 +1954,70 @@ void VoIp::StartCapture(CStdString& party, CStdString& orkuid, CStdString& nativ
...
+	if(orkuid.size())
+	{
+		//VoIpSessionsSingleton::instance()->StartStreamOrkuid(orkuid);
+		LOG4CXX_INFO(s_voipPluginLog, "Undefined");
+	}
+	else if(party.size())
+	{
+		//orkuid = VoIpSessionsSingleton::instance()->StartStream(party);
+		LOG4CXX_INFO(s_voipPluginLog, "Undefined");
+	}
```
Comment (kingster) [inline orkaudio/audiocaptureplugins/voip/VoIp.cpp:1975]: redundant code?

File: orkaudio/audiocaptureplugins/voip/VoIp.cpp [hunk 8]
```diff
@@ -1951,6 +1954,70 @@ void VoIp::StartCapture(CStdString& party, CStdString& orkuid, CStdString& nativ
...
```
Comment (kingster) [inline orkaudio/audiocaptureplugins/voip/VoIp.cpp:2001]: redundant code?

**Review by kingster — CHANGES_REQUESTED** [review]
Few comments
- Please fix code indentation
- Remove non required code flows that do nothing
- Refactor to make this more like a Plugin that can be optionally loaded. No core features should be affected if the plugin is not loaded.
- Other comments are inline.

File: orkaudio/audiocaptureplugins/common/AudioCapturePluginCommon.h [hunk 9]
```diff
@@ -31,6 +31,9 @@ DLL_EXPORT void __CDECL__  Initialize();
 DLL_EXPORT void __CDECL__  Shutdown();
 DLL_EXPORT void __CDECL__  Configure(DOMNode*);
 DLL_EXPORT void __CDECL__  StartCapture(CStdString& party, CStdString& orkuid, CStdString& nativecallid, CStdString& side);
+DLL_EXPORT bool __CDECL__  StartStream(CStdString& nativecallid);
+DLL_EXPORT bool __CDECL__  EndStream(CStdString& nativecallid);
+DLL_EXPORT std::set<std::string> __CDECL__  GetStream();
```
Comment (kingster) [inline orkaudio/audiocaptureplugins/common/AudioCapturePluginCommon.h:36]: `std::*` are not valid C exports

File: orkaudio/audiocaptureplugins/voip/VoIpSession.cpp [hunk 10]
```diff
@@ -4325,6 +4325,16 @@ void VoIpSessions::ClearLocalPartyMap()
...
+std::map<unsigned long long, VoIpSessionRef> VoIpSessions::getByIpAndPort()
+{
+       return m_byIpAndPort;
+}
+
+LoggerPtr VoIpSessions::getLogger()
+{
+       return m_log;
+}
+
```
Comment (kingster) [inline orkaudio/audiocaptureplugins/voip/VoIpSession.cpp:4337]: why does logger needs to exposed outside?

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+static log4cxx::LoggerPtr s_log = log4cxx::Logger::getLogger("plugin.livestream");
+
+template <class T>
+std::string toString(const T &value)
+{
+	std::ostringstream os;
+	os << value;
+	return os.str();
+}
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamFilter.cpp:30]: this function is not used/duplicate. You might want to remove this.

File: orkaudio/filters/LiveStream/LiveStreamServerProxy.cpp [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+#ifndef WIN32
+			LOG4CXX_ERROR(LOG.rootLog, CStdString("Failed to load the following plugin: ") + pluginPath + " " + errstr + CStdString(", could be missing dependency, Try running in cmd box (orkaudio debug)"));
+#else
+			LOG4CXX_ERROR(LOG.rootLog, CStdString("Failed to load the following plugin: ") + pluginPath + CStdString(", could be missing dependency. Try running ldd on it."));
+#endif
+		}
+		else
+		{
+			// Ok, the dll has been successfully loaded
+			LOG4CXX_INFO(LOG.rootLog, CStdString("Loaded plugin: ") + pluginPath);
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamServerProxy.cpp:105]: Why does LiveStream code, load it's on `.so` dynamically? This could be simplified to normal dependency IMO.

File: orkaudio/Makefile.am [hunk 13]
```diff
@@ -6,7 +6,7 @@ AUTOMAKE_OPTIONS = foreign 1.4
 sbin_PROGRAMS = orkaudio
 orkaudio_LDADD =
 orkaudio_LDFLAGS = -lxerces-c -llog4cxx -lorkbase -lsndfile -l@speex_lib@ -lapr-1 -lssl -lcrypto -lpthread -ldl 
-orkaudio_SOURCES = 	OrkAudio.cpp
+orkaudio_SOURCES = 	OrkAudio.cpp filters/LiveStream/LiveStreamServerProxy.cpp filters/LiveStream/LiveStreamServer.cpp
```
Comment (kingster) [inline orkaudio/Makefile.am:9]: Is this actually required? Given that they already being compiled as submodule.

**Review by kingster — COMMENTED** [review]
Looks much better now :)

Please look at the comments.

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+
+	if (isFirstPacket)
+	{
+		headChannel = outputDetails.m_channel;
+		isFirstPacket = false;
+	}
+
+	if (outputDetails.m_channel == headChannel && status)
+	{
+		bufferQueue.push(inputBuffer);
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamFilter.cpp:129]: you might want to check current used size of bufferQueue before pushing into it.

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+			if (srs_audio_write_raw_frame(rtmp, sound_format, sound_rate, sound_size, sound_type, outputBuffer, size, timestamp) != 0)
+			{
+				srs_human_trace("send audio raw data failed.");
+				return;
+			}
+
+			srs_human_trace("Accepted Packet: sample_rate=%d, sound_format=%s, m_timestamp=%d, m_arrivalTimestamp=%d,m_sequenceNumber=%d",
+							outputDetails.m_sampleRate, outputDetails.m_rtpPayloadType, outputDetails.m_timestamp, outputDetails.m_arrivalTimestamp, outputDetails.m_sequenceNumber);
+			srs_human_trace("sent packet: type=%s, time=%d, size=%d, codec=%d, rate=%d, sample=%d, channel=%d",
+							srs_human_flv_tag_type2string(SRS_RTMP_TYPE_AUDIO), timestamp, size, sound_format, sound_rate, sound_size, sound_type);
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamFilter.cpp:155]: move debug logs behind a flag/level.

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+#include "AudioCapture.h"
+#include <iostream>
+#include <string>
+#include <cstring>
+#include "Utils.h"
+#include "srs_librtmp.h"
+#include <queue>
+#include "ConfigManager.h"
+
+#define BUFFER_SAMPLES 8000
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamFilter.cpp:20]: not used

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+	auto key = event->EventTypeToString(event->m_type);
+	LOG4CXX_INFO(s_log, "LiveStream CaptureEventIn " + key + " : " + event->m_value);
+	if (event->m_type == CaptureEvent::EventTypeEnum::EtCallId)
+	{
+		m_callId = event->m_value;
+	}
+
+	if (event->m_type == CaptureEvent::EventTypeEnum::EtKeyValue && event->m_key == "LiveStream" && event->m_value == "start")
+	{
+		std::string url = "rtmp://" + CONFIG.m_rtmpServerEndpoint + ":" + CONFIG.m_rtmpServerPort + "/live/" + m_callId;
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamFilter.cpp:198]: what if `m_callId` is empty?

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    CStdString logMsg;
+    VoIpSessionRef session;
+
+    for (pair = voIpSessions->getByIpAndPort().begin(); pair != voIpSessions->getByIpAndPort().end() && found == false; pair++)
+    {
+        session = pair->second;
+
+        if (session->NativeCallIdMatches(nativecallid))
+        {
+            session->m_keepRtp = true;
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamSession.cpp:46]: What is the effect of this flag?

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+bool LiveStreamSessions::StartStreamNativeCallId(CStdString &nativecallid)
+{
+    std::map<unsigned long long, VoIpSessionRef>::iterator pair;
+    bool found = false;
+    CStdString logMsg;
+    VoIpSessionRef session;
+
+    for (pair = voIpSessions->getByIpAndPort().begin(); pair != voIpSessions->getByIpAndPort().end() && found == false; pair++)
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamSession.cpp:40]: `for ( auto pair ... `

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+#define CONFERENCE_TRANSFER_TRACKING_TAG_KEY "orig-orkuid"
+
+LiveStreamSessions::LiveStreamSessions()
+{
+    voIpSessions = VoIpSessionsSingleton::instance();
+}
+
+bool LiveStreamSessions::StartStreamNativeCallId(CStdString &nativecallid)
+{
+    std::map<unsigned long long, VoIpSessionRef>::iterator pair;
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamSession.cpp:35]: not required to define pair before hand. Pair can be an auto field.

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+    for (pair = voIpSessions->getByIpAndPort().begin(); pair != voIpSessions->getByIpAndPort().end() && found == false; pair++)
+    {
+        session = pair->second;
+
+        if (session->NativeCallIdMatches(nativecallid))
+        {
+            session->m_keepRtp = true;
+            found = true;
+        }
+    }
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamSession.cpp:88]: this is duplicate code in start/stop. You might want to refactor this and create a function

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+    return false;
+}
+
+std::set<std::string> LiveStreamSessions::GetStreamNativeCallId()
+{
+    std::map<unsigned long long, VoIpSessionRef>::iterator pair;
+    bool found = false;
+    CStdString logMsg;
+    VoIpSessionRef session;
+    std::set<std::string> callList;
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamSession.cpp:117]: lot of unused variables

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+
+#include "Utils.h"
+#include "AudioCapture.h"
+#include "LiveStreamSession.h"
+#include "AudioCapturePlugin.h"
+#include "AudioCapturePluginCommon.h"
+#include <set>
+
+extern AudioChunkCallBackFunction g_audioChunkCallBack;
+extern CaptureEventCallBackFunction g_captureEventCallBack;
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamSession.cpp:24]: where are these used?

File: orkaudio/filters/LiveStream/LiveStreamFilter.h [hunk 24]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Oreka -- A media capture and retrieval platform
+ *
+ * G729 Plugin
```
Comment (kingster) [inline orkaudio/filters/LiveStream/LiveStreamFilter.h:4]: Fix plugin name.

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp (hunk 17 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.cpp:178]: @kingster there is a check before calling this function

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp (hunk 14 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.cpp:110]: @kingster Need your help here. How should we decide upon the maxSize for bufferQueue?

File: orkaudio/audiocaptureplugins/common/AudioCapturePluginCommon.h (hunk 9 above)
Comment (shushantsharan) [reply orkaudio/audiocaptureplugins/common/AudioCapturePluginCommon.h]: these changes are non needed anymore

File: orkaudio/audiocaptureplugins/voip/VoIpSession.cpp (hunk 10 above)
Comment (shushantsharan) [reply orkaudio/audiocaptureplugins/voip/VoIpSession.cpp]: removed

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp (hunk 11 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.cpp]: removed

File: orkaudio/filters/LiveStream/LiveStreamServerProxy.cpp (hunk 12 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamServerProxy.cpp]: these changes are non needed anymore

File: orkaudio/Makefile.am (hunk 13 above)
Comment (shushantsharan) [reply orkaudio/Makefile.am]: these changes are non needed anymore

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp (hunk 15 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.cpp]: done

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp (hunk 16 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.cpp]: removed

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp (hunk 19 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamSession.cpp]: done

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp (hunk 20 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamSession.cpp]: done

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp (hunk 21 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamSession.cpp]: done

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp (hunk 22 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamSession.cpp]: refactored

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp (hunk 23 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamSession.cpp]: AudioChunkCallBackFunction is not needed, CaptureEventCallBackFunction is used to push start/stop events to the filter

File: orkaudio/filters/LiveStream/LiveStreamFilter.h (hunk 24 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.h]: done

File: orkaudio/filters/LiveStream/LiveStreamSession.cpp (hunk 18 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamSession.cpp]: copied by mistake

**Comment by shushantsharan** [comment]
@kingster **Latest Structure**:
Files added : 
- srs_librtmp.cpp, srs_librtmp.h --> RTMP client Library 
- LiveStreamFilter.cpp, LiveStreamFilter.h --> Rtp packets flow through this filter and are ingested to rtmp server from here ondemand
- LiveStreamSession.cpp, LiveStreamSession.h --> finds session for a nativeCallId and pushed start/stop event to the session
- LiveStreamServer.cpp, LiveStreamServer.h, json.h --> HTTP server

POP based refactoring key updates : 
- All code changes added to **orkaudio.cpp** are removed now.
- LiveStreamServer is started when _OrkInitialize()_ is invoked for _LiveStreamFilter_ . This helps in eliminating almost all the code changes from the rest of the project

File: orkaudio/filters/LiveStream/json.h [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+#define JSON_HEADER_FILE_H
+
+/**
+* Need to undef snprintf other MSVC throws error that it is not part of std::
+* Somewhere when including pjsua2.hpp this gets set.
+*/
+#undef snprintf
+#include <nlohmann/json.hpp>
+
+#endif
```
Comment (kingster) [inline orkaudio/filters/LiveStream/json.h:13]: You can skip this file completely. Just include `#include <nlohmann/json.hpp>` directly. This isn't a MSVC app, hence you can skip the logic of `snprintf`.

File: orkaudio/OrkAudio.cpp [hunk 26]
```diff
@@ -401,27 +394,10 @@ void MainThread()
...
-
-	// get void*'s for all entries on the stack
-	size = backtrace(array, 10);
-
-	// print out all the frames to stderr
-	fprintf(stderr, "Error: signal %d:\n", sig);
-	backtrace_symbols_fd(array, size, STDERR_FILENO);
-	#endif
-  	exit(1);
-}
```
Comment (kingster) [inline orkaudio/OrkAudio.cpp:418]: can you please revert this change? This is for backtrace if the app crashes.

File: orkaudio/filters/LiveStream/LiveStream.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStream.cpp]: fixed

File: orkaudio/audiocaptureplugins/voip/VoIpSession.cpp (hunk 5 above)
Comment (shushantsharan) [reply orkaudio/audiocaptureplugins/voip/VoIpSession.cpp]: done

File: orkaudio/filters/LiveStream/LiveStream.h (hunk 6 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStream.h]: done

File: orkaudio/audiocaptureplugins/voip/VoIp.cpp (hunk 7 above)
Comment (shushantsharan) [reply orkaudio/audiocaptureplugins/voip/VoIp.cpp]: removed

File: orkaudio/audiocaptureplugins/voip/VoIp.cpp (hunk 8 above)
Comment (shushantsharan) [reply orkaudio/audiocaptureplugins/voip/VoIp.cpp]: removed

File: orkaudio/filters/LiveStream/LiveStreamFilter.cpp (hunk 14 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStreamFilter.cpp]: fixed

File: orkaudio/filters/LiveStream/json.h (hunk 25 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/json.h]: removed

File: orkaudio/filters/LiveStream/LiveStream.h (hunk 3 above)
Comment (shushantsharan) [reply orkaudio/filters/LiveStream/LiveStream.h]: done

**Review by kingster — APPROVED** [review]

**Comment by vcidst** [comment]
Hey, this seems interesting. Did you folks make any headway into this? I've been looking for ways to connect SIP calls to Dialogflow framework.

**Comment by kingster** [comment]
Hi @vcidst 

These changes are stable right now, and we have been using this feature in our production for quite some time, with the only caveat being that only G711 (PCMA/PCMU) is supported. 

You can try out with the [builds from this pr](https://github.com/voiceip/oreka/actions/runs/2206447577), and the [wiki for setup instructions](https://github.com/voiceip/oreka/wiki/Live-Streaming)
